File: customer_journey/customer_journey/page/customer_safari/customer_safari.js
var cust_name
var counter = 0
frappe.pages['customer_safari'].on_page_load = function(wrapper) {
        var page = frappe.ui.make_app_page({
                parent: wrapper,
                title: 'Customer-360°',
                single_column: true
        });

        frappe.show_alert('Hey '+frappe.session.user,3)

        page.custom_field = page.add_field({
                fieldname: 'customer',
                label: __('Customer'),
                fieldtype: 'Link',
                options: 'Customer',
                change: function(){
			 
                        if(counter==0){
console.log(page.fields_dict.customer.value)
				console.log("suppose to have a customer")
                               cust_name = page.fields_dict.customer.value
                              abc()
                              console.log("here I am")
                                counter++

                        }else{
                                counter = 0;
                                return false;
                        }

                }
        });

       



 frappe.breadcrumbs.add("CRM");

}





function abc(){
console.log("line232", cust_name)
        if(cust_name==undefined){
            console.log("undefined", cust_name)
                frappe.show_alert("Please Select Customer First")
        }else{
        var a = $(".page-wrapper").find("#nik")
console.log("aa",a)
        if(a.length === 0){
console.log("not zero")
        }else{
                $("#nik").remove()
        }


        frappe.call({
                method: 'customer_journey.customer_journey.page.customer_safari.customer_safari.customer_details',
                args: {
                        name:cust_name
                },
                callback:function(r){
console.log(cust_name)
                         console.log("r",r)
			if(r.message==undefined || r.message=="" || r.message == "No_leads"){
				frappe.show_alert("This customer does'nt have a lead Please Select Another Customer")
			}else{

//Rendering the html page
                         var html = r.message['Template']
                         $(".page-wrapper").append("<div id=nik></div>")

                         big_object1 = r.message
                         creation_data = big_object1["Creation_data"]
                         console.log(creation_data)
                         var rpm_days = null_check(creation_data["RPM_diff"])
                         var opportunity_days =  null_check(creation_data["Opportunity_diff"])
                         var kyc_days =  null_check(creation_data["KYC_diff"])
                         var customer_days =  null_check(creation_data["Customer_diff"])
                         var payment_days =  null_check(creation_data["Payment_diff"])
                         var Invoice_days =  null_check(creation_data["Invoice_diff"])
                        
                         var rpm_creation =  null_check(creation_data["RPM_creation"])
                         var opportunity_creation =  null_check(creation_data["Opportunity_creation"])
                         var kyc_creation =  null_check(creation_data["KYC_creation"])
                         var customer_creation =  null_check(creation_data["Customer_creation"])
                         var payment_creation =  null_check(creation_data["Payment_creation"])
                         var invoice_creation =  null_check(creation_data["Invoice_creation"])  
                         var upgrade_creation = null_check(creation_data["Upgrade_creation"])   
                         var sovp_creation = null_check(creation_data["SOVP_creation"])
                         var sovp_days = null_check(creation_data["SOVP_diff"])
                         var lead_creation = null_check(creation_data["Lead_creation"])
                         
                         var upgrade_days = null_check(creation_data["Upgrade_diff"])
                         var rpm2_days = null_check(null)
                         var rpm2_creation = null_check(null)
                         var product_sell_days = null_check(null)
                         var product_sell_creation = null_check(null)
                         var renewal_days = null_check(null)
                         var renewal_creation = null_check(null)
                         var service_expired_days = null_check(null)
                         var service_expired_creation = null_check(null) 
                       
                         html = html.replace("{{RPM_days}}", rpm_days)
                         html = html.replace("{{Opportunity_days}}", opportunity_days)
                         html = html.replace("{{KYC_days}}", kyc_days)
                         html = html.replace("{{Customer_days}}", customer_days)
                         html = html.replace("{{Payment_days}}", payment_days)                         
                         html = html.replace("{{Invoice_days}}", Invoice_days)
                         html = html.replace("{{SOVP_days}}", sovp_days)
                         html = html.replace("{{Lead_creation}}", lead_creation)
                         html = html.replace("{{Upgrade_days}}", upgrade_days)
                         html = html.replace("{{Upgrade_creation}}", upgrade_creation)
                         html = html.replace("{{RPM_creation}}", rpm_creation)
                         html = html.replace("{{Opportunity_creation}}", opportunity_creation)
                         html = html.replace("{{KYC_creation}}", kyc_creation)
                         html = html.replace("{{Customer_creation}}", customer_creation)
                         html = html.replace("{{Payment_creation}}", payment_creation)                         
                         html = html.replace("{{Invoice_creation}}", invoice_creation)
                         html = html.replace("{{SOVP_creation}}", sovp_creation) 
                         html = html.replace("{{RPM2_creation}}", rpm2_creation)
                         html = html.replace("{{RPM2_days}}", rpm2_days)
                         html = html.replace("{{Product Sell_creation}}", product_sell_creation)
                         html = html.replace("{{Product Sell_days}}", product_sell_days)
                         html = html.replace("{{Renewal_creation}}", renewal_creation)
                         html = html.replace("{{Renewal_days}}", renewal_days)
                         html = html.replace("{{Service Expired_creation}}", service_expired_creation)
                         html = html.replace("{{Service Expired_days}}", service_expired_days)
                       


 
                         $("#nik").append(html)

                         var counter =0
                         var big_object = r.message
                         console.log(big_object)  
                         var element_list = ["Lead" ,"RPM" ,"Opportunity" ,"KYC" , "Customer" ,"Payment Entry" ,"Sales Invoice" ,"SVP" ,"Upgrade"]
                         
                         for(var j in element_list){
                                   
                                   let div = document.createElement("div");
                                   div.setAttribute('class', 'arrow_box-'+ counter)
                                   
                                   if(big_object[element_list[j]] != undefined &&  Object.keys(big_object[element_list[j]]).length > 0){
                                      
                                       if(big_object[element_list[j]]["type"] == "nested"){
                                           let key_list2 = Object.keys(big_object[element_list[j]])
                                           delete big_object[element_list[j]].type
                                           let li =  document.createElement("li");
                                           li.innerText  = "Count: " +  big_object[element_list[j]].Count
                                           div.appendChild(li)
                                           
                                           for(var key in big_object[element_list[j]]){
                                               
                                               
				               var element_object = big_object[element_list[j]][key]   
                                               var key_list = Object.keys(element_object)
                                               if (key_list!=undefined){
                                                    for (var i in key_list){
                                                        
                                                       
                                                        let field = document.createElement("p")
                                                        let link = document.createElement("a");
                                                        link.innerText = key + " link"
                                                        let brk = document.createElement("br")
                                                        link.href =  "desk#Form/" + element_list[j] +"/" + element_object[key_list[i]]
				                        link.target = "_blank"	
                                                        
                                                        div.appendChild(link)
                                                        div.appendChild(brk)
                                                        
                                                        document.getElementById(element_list[j]).appendChild(div);
                                                    }
                                        
                                               }
                                           }
                                       }
                                       else{
//                                           console.log(big_object[element_list[j]].length)  
                                           var element_obj = big_object[element_list[j]]
                                           var key_list = Object.keys(element_obj)
                                           if (key_list!=undefined){
                                               for (var i in key_list){
                                                   
                                                   let li =  document.createElement("li");
                                                   let field = document.createElement("p")
				                   field.innerText = key_list[i]
                                                   field.setAttribute('style', 'font-weight: bold')
                                                   li.innerText = field.innerText +": " +element_obj[key_list[i]] 
                                                   div.appendChild(li)
                                                   console.log(element_list[j])
                                                   document.getElementById(element_list[j]).appendChild(div);
                                               }
                                           }
                                       }
                                   }
                                   else{
                                       let li =  document.createElement("li");
                                       li.innerText = element_list[j] + " data not available"
                                       div.appendChild(li)
                                       console.log(element_list[j])
                                       document.getElementById(element_list[j]).appendChild(div);
                                       }
                                   console.log(counter)
                                   counter += 1
                         }
                         function null_check(field){
			                            if (field==undefined || field==null){
                                                         return "not available"
                                                        }
                                                    else{
                                                        return field                 
                                                        }
                                                    }         

		         var lead = r.message['Lead Id']
                         var mail = r.message['Email']
                         var mobile = r.message['Primary Mobile']
                         var name = r.message['Customer Name']
                         var creation = moment(r.message['Creation']).format("DD-MMM-YYYY HH:mm A")
                         var cust_creat =  moment(r.message['Customer Creation']).format("DD-MMM-YYYY HH:mm A")
  //                       var addr = r.message['Address']
                         var cust_status = r.message['Current Status']
                         var profile = r.message['Image']
			 var campaign_name = r.message['Campaign Name']
			 var lead_src = r.message['Lead Source']
			var test = r.message['test']
			 var alter_mob = r.message['Alternate Number']
			 var oth_mob = r.message['Other Number']
			 var risk = r.message['Risk Appetite']
			 var rpm_by = r.message['RPM By']
			 var rpm_on = moment(r.message['RPM On']).format("DD-MMM-YYYY HH:mm A")
			 var sales_person = r.message['Sales Person']
			 var opportunity = moment(r.message['Opportunity']).format("DD-MMM-YYYY HH:mm A")
			 var kyc = r.message['KYC']
			 var sms = r.message['Total SMS']
			 var tot_mail = r.message['Total Email']
			 var pay_entry_date = moment(r.message['Payment Entry Date']).format("DD-MMM-YYYY HH:mm A")
			 var pay_entry_amt = r.message['Payment Entry Amount']
			 var invoice = r.message['Invoice']
			 var upgrade = r.message['Upgrade']
			 var rpm_dif = r.message['RPM Diff']
			 var opp_dif = r.message['Opp Diff']
			 var kyc_dif = r.message['KYC Diff']
			 var cust_dif = r.message['Cust Diff']
			 var pay_dif  = r.message['Pay Diff']
			 var inv_dif = r.message['Inv Diff']
			 var sovp_diff = r.message['SOVP Diff']
			 var cust_app = moment(r.message['Customer Approved']).format("DD-MMM-YYYY HH:mm A")
			 var bank_acc = r.message['Bank Account']
			 var currency = r.message['Currency']

                        

//			
//			var inv_code=""
//			var sovp_code=""
//			
//				for(var i=0;i<invoice.length;i++){
//				
//				inv_code = inv_code+invoice[i]['name']+" "+invoice[i]['sales_person']+" "+invoice[i]['items'][0]['item_code']+"<hr/>"
//
//				sovp_code = sovp_code + invoice[i]['items'][0]['item_code']+" "+invoice[i]['workflow_state']+" "+invoice[i]['from_date']+" "+invoice[i]['end_date']+"<hr/>"
//				}	
	
			
			


		
			
	
			//Hiding 
		/*
			if(cust_status!='Approved'){
				html = html.replace("{{upgrade_hide}}","<!--")
				html = html.replace("{{service_expired_hide}}","-->")
			}else if(invoice.length==0 || invoice=="" || invoice==undefined){
				 html = html.replace("{{upgrade_hide}}","<!--")
                                html = html.replace("{{service_expired_hide}}","-->")	
			}else if(pay_entry_date==undefined ||  pay_entry_date==""){
				 html = html.replace("{{upgrade_hide}}","<!--")
                                html = html.replace("{{service_expired_hide}}","-->")
			}
			
			if(upgrade!='Yes'){

				console.log('inside upgrade function')
				html = html.replace("{{upgrade_hide}}", $('#upgrade_hide').blur())
				html = html.replace("{{upgrade_hide_1}}",-->)
				html = html.replace("{{rpm_2}}",<!--)
				html = html.replace("{{service_expired_hide}}",-->)
			}
				*/


                        
 


			//Pushing Values
//		
//			html = html.replace("{{Bank Account}}",bank_acc);
//			html = html.replace("{{RPM On}}",rpm_on);
//			html = html.replace("{{RPM By}}",rpm_by);
//			html = html.replace("{{SOVP_INV}}",sovp_code);
//			html = html.replace("{{INV}}",inv_code);
//	                html = html.replace("{{Lead ID}}",lead);
//			html = html.replace("{{Email}}",mail);
//			html = html.replace("{{Customer Approved}}",cust_app);
//			html = html.replace("{{SOVP Diff}}",sovp_diff);
//			html = html.replace("{{Inv Diff}}",inv_dif);
//			html = html.replace("{{Customer Approved}}",cust_app);
//			html = html.replace("{{Pay Diff}}",pay_dif);
//			html = html.replace("{{Cust Diff}}",cust_dif);
//			html = html.replace("{{KYC Diff}}",kyc_dif);
//			html = html.replace("{{Opp Diff}}",opp_dif);
//			html = html.replace("{{RPM Diff}}",rpm_dif);
//			html = html.replace("{{Upgrade}}",upgrade);
//			html = html.replace("{{Payment Entry Amount}}",pay_entry_amt +" "+ currency);
//			html = html.replace("{{Payment Entry Date}}",pay_entry_date);
//			html = html.replace("{{Payment Entry Date}}",pay_entry_date);
//			html = html.replace("{{Total Email}}",tot_mail);
//			html = html.replace("{{Total SMS}}",sms);
//			html = html.replace("{{KYC}}",kyc);
//			html = html.replace("{{Sales Person}}",sales_person);
//			html = html.replace("{{Opportunity}}",opportunity);
//			html = html.replace("{{RPM On}}",rpm_on);
//			html = html.replace("{{RPM By}}",rpm_by);
//			html = html.replace("{{Risk Appetite}}",risk);
//			html = html.replace("{{Other Number}}",oth_mob);
//                        html = html.replace("{{Alternate Number}}",alter_mob);
//			html = html.replace("{{Lead Source}}",lead_src);
//                        html = html.replace("{{Campaign Name}}",campaign_name);
//			html = html.replace("{{test}}", test[0]);
//			html = html.replace("{{Image}}",profile);
//                        html = html.replace("{{Current Status}}",cust_status);
////			html = html.replace("{{Address}}",addr);
//			html = html.replace("{{Customer Creation}}",cust_creat);
//			html = html.replace("{{Lead Creation}}",creation);
//		        html = html.replace("{{Customer Name}}",name);
//			html = html.replace("{{Primary Mobile}}",mobile);
//			html = html.replace("{{Lead Creation}}",creation);
//		//	html = html.replace("{{Opportunity}}",opportunity);


               
	
                //      $("#nik").append("<h5>Lead ID: "+lead+"<br>E-Mail: "+mail+"<br>Name: "+name+"<br>Gender: "+gender+"<br>Mobile: "+mobile+"<br>First Onboarded On$
	
                }
		}
        });
        }
}

File: customer_journey/customer_journey/page/customer_safari/customer_safari.py

import frappe
from datetime import datetime
from collections import OrderedDict

def days_between(d1, d2):
    if type(d1) == type(d2):
        if(d1 > d2):
            return (d1 - d2).days
        else:
    	    return (d2 - d1).days
    else:
	return "not defined"

def handle_null_values(field):
	if not field:
	     return "not available"
        else:
	     return field

    
def prepare_data_dict(customer,lead, risk_date, opportunity, html, pay, invoice, email_count, campaign):
    data_dict = {}
    inv_data = []
    data_dict["Creation_data"] = {} 

    if (len(risk_date)>0 and risk_date[0][1] is not None) and (lead[0] is not None and lead[0][2] is not None):
        data_dict["Creation_data"]["Lead_creation"] = lead[0][2].strftime("%d-%m-%Y")
        data_dict["Creation_data"]["RPM_diff"] = days_between(lead[0][2], risk_date[0][1]) #difference between lead creation and firt RPM creation
        data_dict["Creation_data"]["RPM_creation"] = risk_date[0][1].strftime("%d-%m-%Y")
    
    if len(opportunity)>0: 
        if (len(opportunity[0])>0 and opportunity[0][0] is not None) and (lead[0] is not None and lead[0][2] is not None):
	    data_dict["Creation_data"]["Opportunity_diff"] =  days_between(lead[0][2], opportunity[0][0]) #difference between lead creation and firt opportunity creation
	    data_dict["Creation_data"]["Opportunity_creation"] = opportunity[0][0].strftime("%d-%m-%Y")
   
    if (customer[0][2] is not None and customer[0][5] is not None):
        data_dict["Creation_data"]["KYC_creation"] = customer[0][5].strftime("%d-%m-%Y") 
        data_dict["Creation_data"]["KYC_diff"] = days_between(customer[0][2], customer[0][5]) #difference between customer creation and customer approced_on

    if (lead[0][2] is not None and customer[0][2] is not None):
        data_dict["Creation_data"]["Customer_diff"] = days_between(lead[0][2], customer[0][2]) #difference between lead creation and firt RPM creation
        data_dict["Creation_data"]["Customer_creation"] = customer[0][2].strftime("%d-%m-%Y")
    if pay:
        if pay[0][0] and customer[0][2] is not None:
           data_dict["Creation_data"]["Payment_diff"] =  days_between(customer[0][2], pay[0][0]) #difference between customer creation and firt payment creation
	   data_dict["Creation_data"]["Payment_creation"] = pay[0][0].strftime("%d-%m-%Y")
	    
    data_dict["SVP"] = OrderedDict()       
    data_dict["Upgrade"] =  OrderedDict()
    data_dict["RPM"] =  OrderedDict()
    data_dict["Opportunity"] =  OrderedDict()
    data_dict["Customer"] =  OrderedDict()
    
  
# adding customer details to data_dict
    if customer[0] is not None: 
        customer_detail_list = customer[0]
       
        data_dict["KYC"] = OrderedDict() 
        data_dict["Customer"]["Name"] = customer_detail_list[0] 
	if customer_detail_list[4]:
	    if customer_detail_list[4] == "Approved":
	        data_dict["KYC"]["Status"] = "Approved"
	    else:
                data_dict["KYC"]["Status"] = "Pending"
        else:
    	    data_dict["KYC"]["Status"] = "KYC status is not available"
    if email_count:
        data_dict["Customer"]["Emails"] = email_count
## adding lead details to datat_dict 

    if lead[0]:
        data_dict["Lead"] = OrderedDict()
	lead_detail_list = lead[0]
        if campaign and campaign[0]:
            campaign_detail_list = campaign[0]
            
            if campaign_detail_list[1]:
	        data_dict["Lead"]["Source"] =  campaign_detail_list[1]
	    else:
	        data_dict["Lead"]["Source"] = "Lead source not define"
            if campaign_detail_list[2]:
               data_dict["Lead"]["Medium"] = campaign_detail_list[2]
	if lead_detail_list[5]:
            data_dict["Lead"]["Campaign"] = lead_detail_list[5]
        else:
            data_dict["Lead"]["Campaign"] = "Lead Id not available"
        if lead_detail_list[9]:
            data_dict["RPM"]["Risk Apptite"] = lead_detail_list[9]
        else:  
            data_dict["RPM"]["Risk Apptite"] = "not available"
## calculation for total SMS
        
        prim_mob = lead_detail_list[3] 
        sec_mob =  lead_detail_list[7] 
        oth_mob =  lead_detail_list[8]
        
        prim_sms = frappe.db.count('SMS Log', filters={'sent_to': ('like', '%'+prim_mob+'%')}) if prim_mob is not None else '0'
        sec_sms = frappe.db.count('SMS Log', filters={'sent_to': ('like', '%'+sec_mob+'%')}) if sec_mob is not None else '0'
        oth_sms = frappe.db.count('SMS Log',filters={'sent_to': ('like', '%'+oth_mob+'%')}) if oth_mob is not None else '0'

        total_sms = int(prim_sms) + int(sec_sms) + int(oth_sms)
        data_dict["Customer"]["SMS"] = total_sms    

# adding opportunity details to data_dict    
    if opportunity:
        
        data_dict["Opportunity"]["type"] = "nested"
        data_dict["Opportunity"]["Count"] = len(opportunity) 
        for index, val in enumerate(opportunity):
              opportunity_detail_list = val
              data_dict["Opportunity"][opportunity_detail_list[3]] = {}   
	     
              
	      if opportunity_detail_list[3]:
	          data_dict["Opportunity"][opportunity_detail_list[3]]["id"] = opportunity_detail_list[3]
	      else:
	          data_dict["Opportunity"][opportunity_detail_list[3]]["id"] = "not available"
	     
    if pay:
        data_dict["Payment Entry"] = {} 
        data_dict["Payment Entry"]["type"] = "nested"
        data_dict["Payment Entry"]["Count"] = len(pay)
        for index, val in enumerate(pay):
            pay_detail_list = val
            data_dict["Payment Entry"][pay_detail_list[4]] = {}
	    if pay_detail_list[4]:
	         data_dict["Payment Entry"][pay_detail_list[4]]["id"] = pay_detail_list[4]
	    else:
	         data_dict["Payment Entry"][pay_detail_list[4]]["id"] = "Payment entry date is not available"
    data_dict["Testing"] = {}
    	     
    if invoice: 
        for i in invoice:
	     a = frappe.get_doc("Sales Invoice",i['name'])
	     inv_data.append(a)
        invoice_date = []
	product = set()
        tenure = set()	    	    
	counter = 0
        if len(inv_data)>0:
            for j in range(0,len(inv_data)):
                items = inv_data[j].get("items")
                item_name = items[0].get("item_name")
                price_list = inv_data[j].get("selling_price_list")
                if item_name in product and price_list not in tenure:
                    counter=counter+1
                    invoice_date.append(inv_data[j].get("creation"))
                else:
                    if item_name:
                        product.add(item_name)
                    if price_list:
                        tenure.add(price_list)
        data_dict["Testing"]["product"] = product
        data_dict["Testing"]["item"] = invoice_date
	if counter>0:
            upgradation='Yes'
        else:
            upgradation='No'
    else:
         upgradation = "No Invoice record found"
   
    if len(inv_data)>0  and customer[0][0] is not None:
       
       data_dict["Creation_data"]["Invoice_diff"] =  days_between(customer[0][2], inv_data[0].get("creation")) #difference between customer creation and firt payment creation
       data_dict["Creation_data"]["Invoice_creation"] = inv_data[0].get("creation").strftime("%d-%m-%Y")

       if inv_data[0].get("approved_on"):
           
           data_dict["Creation_data"]["SOVP_creation"] = inv_data[0].get("approved_on").strftime("%d-%m-%Y")  
           data_dict["Creation_data"]["SOVP_diff"] = days_between(inv_data[0].get("creation"), inv_data[0].get("approved_on"))

           if upgradation == "Yes": 
               data_dict["Creation_data"]["Upgrade_creation"] = inv_data[0].get("approved_on").strftime("%d-%m-%Y")
              # data_dict["Creation_data"]["Upgrade_diff"] = days_between(invoice_date[0], invoice_date[1])          
        
    if len(inv_data)>0:
       data_dict["Sales Invoice"] = {}
       data_dict["Sales Invoice"]["type"] = "nested"
       data_dict["Sales Invoice"]["Count"] = len(inv_data)
       
       for index, item in enumerate(inv_data):
           data_dict["Sales Invoice"][item.get("name")] = {}
	   data_dict["Sales Invoice"][item.get("name")]["id"] = item.get("name")
	  
    
    
    data_dict["Upgrade"]["Is_upgraded"] = upgradation 
    
    data_dict["Template"] = html.template
    return data_dict	  
   	

   
	
@frappe.whitelist()
def customer_details(name):
        
    customer = frappe.db.sql("""select customer_name,lead_name,creation,docstatus,workflow_state,approved_on from `tabCustomer` where name=%s """,(name)) 
     	
    lead_id = customer[0][1]
    if not lead_id:
        return "No_leads"
    lead = frappe.db.sql("""select name,email_id,creation,primary_mobile,image,campaign_name,source,alternate_phone,other_number,risk_appetite,sales_person from `tabLead` where name=%s """,(lead_id))
    if not lead:
	return "No_leads"
 
   
 	
    if lead[0][5]:
        campaign_name = lead[0][5]
        campaign = frappe.db.sql("""select campaign_name, source, medium from `tabCampaign` where name=%s """,(campaign_name)) 

    risk_date = frappe.db.sql(""" select owner,creation from `tabRisk Profile` where  document_number=%s  """,(lead_id))

    opportunity = frappe.db.sql(""" select creation,sales_person,contact_by, name  from tabOpportunity where lead=%s order by creation  """,(lead_id))
   	
   
    html = frappe.get_doc('Safari Template','Safari Template')

    
    email_count = frappe.db.count("Email Queue Recipient",filters={'recipient':('like','%'+lead[0][1]+'%')})
    	
    pay = frappe.db.sql("""select creation,paid_amount,bank_account,paid_to_account_currency, name from `tabPayment Entry` where title=%s order by creation  """,(name))
       
    
         

    invoice = frappe.get_list("Sales Invoice", filters = {'customer': name}, fields = ['name'], order_by = 'creation')
    
    
    data_dict = prepare_data_dict(customer , lead, risk_date, opportunity, html, pay, invoice, email_count, campaign)  
    return data_dict 
